src/main/java/org/skife/jdbi/v2/sqlobject/stringtemplate/UseStringTemplate3StatementLocator.java

```diff
@@ -37,13 +41,32 @@
 
     String value() default DEFAULT_VALUE;
     Class errorListener() default StringTemplateErrorListener.class;
+    boolean cacheable() default true;
 
     class LocatorFactory implements SqlStatementCustomizerFactory
     {
+        @GuardedBy("LocatorFactory.class")
+        private static final Map<Class<?>, WeakHashMap<Annotation, SqlStatementCustomizer>> CUSTOMIZER_CACHE =
+                new WeakHashMap<Class<?>, WeakHashMap<Annotation, SqlStatementCustomizer>>();
+
         @Override
         public SqlStatementCustomizer createForType(Annotation annotation, Class sqlObjectType)
         {
             final UseStringTemplate3StatementLocator a = (UseStringTemplate3StatementLocator) annotation;
+
+            if (a.cacheable()) {
+                synchronized (LocatorFactory.class) {
```

Comment:
In fact I'd say don't lock here at all, and move the `CUSTOMIZER_CACHE.put` call down at the bottom of the method, where you can lock unconditionally. That way the top section that checks the cache is a read-only operation and therefore safe to access without synchronization.